locator/tablet_sharder.hh

```diff
@@ -28,6 +28,68 @@ private:
             _tmap = &_tm.tablets().get_tablet_map(_table);
         }
     }
+
+    std::optional<unsigned> get_shard(const tablet_replica_set& replicas, host_id host) const {
+        for (auto&& r : replicas) {
+            if (r.host == host) {
+                return r.shard;
+            }
+        }
```

Comment:
Even the common cases here seem to be slow and we should cache them.

Reply:
But, not urgent. The series should contain a perf-simple-query report (for the common case of no migration).

Reply:
shouldn't we on_internal_error if no hosts? wouldn't a read be fenced off?

Reply:
Good question. 

I chose to preserve the existing contract of the sharder::shard_of() API, which is to provide a shard for any token, to avoid surprises. If we throw, some internal algorithms which walk the whole ring could get broken. storage_proxy requests would be fine with throwing because they should not route to non-owners, but other internal uses may not be fine with it. The way things currently are, replica layer is unaware of ownership. You can read any token even if the node (not just shard) doesn't own it. Ownership is managed at storage_proxy level. We could consider changing that, but I'd rather do it later.

Reply:
Keeping the status quo is okay, but I want to eventually move away from it. It can easily hide bugs. Please add a FIXME.

Reply:
> The series should contain a perf-simple-query report (for the common case of no migration).

Report added to the covering letter. No sign of regression. Difference within noise.

Reply:
> Keeping the status quo is okay, but I want to eventually move away from it. It can easily hide bugs. Please add a FIXME.

```patch
--- a/locator/tablet_sharder.hh
+++ b/locator/tablet_sharder.hh
@@ -110,6 +110,9 @@ class tablet_sharder : public dht::sharder {
     virtual unsigned shard_for_reads(const token& t) const override {
         ensure_tablet_map();
         auto tid = _tmap->get_tablet_id(t);
+        // FIXME: Consider throwing when there is no owning shard on the current host rather than returning 0.
+        // It's a coordination mistake to route requests to non-owners. Topology coordinator should synchronize
+        // with request coordinators before moving the shard away.
         auto shard = shard_for_reads(tid, _host).value_or(0);
         tablet_logger.trace("[{}] shard_of({}) = {}, tablet={}", _table, t, shard, tid);
         return shard;
```